concrete/src/Http/Middleware/ThumbnailMiddleware.php

```diff
@@ -86,32 +92,77 @@ public function process(Request $request, DelegateInterface $frame)
     }
 
     /**
+     * @param Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    private function buildRedirectToThumbnailResponse(Request $request)
+    {
+        $redirectTo = (string) $request->getUri();
+        $redirectTo .= ((strpos($redirectTo, '?') === false) ? '?' : '&') . mt_rand();
+        $responseFactory = $this->app->make(ResponseFactoryInterface::class);
+        $response = $responseFactory->redirect($redirectTo);
+
+        return $response;
+    }
+
+    /**
+     * @param Connection $database
+     * @param string|null $searchThumbnailPath
+     *
+     * @return ConcretePDOStatement|null
+     */
+    private function getThumbnailPathsToGenerate(Connection $database, $searchThumbnailPath = null)
+    {
+        if ($searchThumbnailPath === null) {
+            $rs = $database->executeQuery('SELECT * FROM FileImageThumbnailPaths WHERE isBuilt = 0 LIMIT 5');
+        } else {
+            $rs = $database->executeQuery('SELECT * FROM FileImageThumbnailPaths WHERE isBuilt = 0 ORDER BY ' . $database->getDatabasePlatform()->getLocateExpression('?', 'path') . ' DESC LIMIT 5', [$searchThumbnailPath]);
+        }
+        if (!$rs->rowCount()) {
+            $rs->closeCursor();
+            $rs = null;
+        }
+
+        return $rs;
+    }
+
+    /**
      * Generate thumbnails and and manage db update.
      *
-     * @param array[]                                       $paths
+     * @param ConcretePDOStatement $paths
      * @param \Concrete\Core\Database\Connection\Connection $database
+     * @param string|null $searchThumbnailPath
+     *
+     * @return bool returns true if $searchThumbnailPath is set and one of the thumbnails may be for that thumbnail
      */
-    private function generateThumbnails($paths, Connection $database)
+    private function generateThumbnails(ConcretePDOStatement $paths, Connection $database, $searchThumbnailPath)
     {
-        $database->transactional(function (Connection $database) use ($paths) {
+        $result = false;
+        $database->transactional(function (Connection $database) use ($paths, $searchThumbnailPath, &$result) {
             foreach ($paths as $thumbnail) {
                 /** @var File $file */
                 $file = $this->getEntityManager()->find(File::class, $thumbnail['fileID']);
 
                 if ($this->attemptBuild($file, $thumbnail)) {
                     $this->completeBuild($file, $thumbnail);
+                    if ($result === false && $this->maybeTheRequestedThumbnail($thumbnail, $searchThumbnailPath)) {
+                        $result = true;
```

Comment:
This is confusing to me, how is this `$result` related to generating the thumbnail? I would just loop over the paths again after generating thumbnails completes.